extensions/typescript-language-features/src/languageFeatures/inlayHints.ts

```diff
@@ -0,0 +1,143 @@
+
+import * as vscode from 'vscode';
+import * as Proto from '../protocol';
+import { DocumentSelector } from '../utils/documentSelector';
+import { ClientCapability, ITypeScriptServiceClient, ServerResponse, ExecConfig } from '../typescriptService';
+import { conditionalRegistration, requireMinVersion, requireSomeCapability } from '../utils/dependentRegistration';
+import { Position } from '../utils/typeConverters';
+import FileConfigurationManager, { getInlayHintsPreferences } from './fileConfigurationManager';
+import API from '../utils/api';
+import { isTypeScriptDocument } from '../utils/languageModeIds';
+
+namespace ExperimentalProto {
+	export const enum CommandTypes {
+		ProvideInlineHints = 'ProvideInlayHints'
+	}
+
+	export interface InlayHintsArgs extends Proto.FileRequestArgs {
+		/**
+		 * Start position of the span.
+		 */
+		start: number;
+		/**
+		 * Length of the span.
+		 */
+		length: number;
+	}
+
+	export interface InlineHintsRequest extends Proto.Request {
+		command: CommandTypes.ProvideInlineHints;
+		arguments: InlayHintsArgs;
+	}
+
+	export enum InlayHintKind {
+		Type = 'Type',
+		Parameter = 'Parameter',
+		Enum = 'Enum'
+	}
+
+	interface InlayHintItem {
+		text: string;
+		position: Proto.Location;
+		kind?: InlayHintKind;
+		whitespaceBefore?: boolean;
+		whitespaceAfter?: boolean;
+	}
+
+	export interface InlayHintsResponse extends Proto.Response {
+		body?: InlayHintItem[];
+	}
+
+	export interface IExtendedTypeScriptServiceClient {
+		execute<K extends keyof ExtendedTsServerRequests>(
+			command: K,
+			args: ExtendedTsServerRequests[K][0],
+			token: vscode.CancellationToken,
+			config?: ExecConfig
+		): Promise<ServerResponse.Response<ExtendedTsServerRequests[K][1]>>;
+	}
+
+	export interface ExtendedTsServerRequests {
+		'provideInlayHints': [InlayHintsArgs, InlayHintsResponse];
+	}
+
+	export namespace InlayHintKind {
+		export function fromProtocolInlayHintKind(kind: InlayHintKind): vscode.InlayHintKind {
+			switch (kind) {
+				case InlayHintKind.Parameter:
+					return vscode.InlayHintKind.Parameter;
+				case InlayHintKind.Type:
+					return vscode.InlayHintKind.Type;
+				case InlayHintKind.Enum:
+					return vscode.InlayHintKind.Other;
+				default:
+					return vscode.InlayHintKind.Other;
+			}
+		}
+	}
+}
+
+class TypeScriptInlayHintsProvider implements vscode.InlayHintsProvider {
+	public static readonly minVersion = API.v440;
+
+	constructor(
+		private readonly client: ITypeScriptServiceClient,
+		private readonly fileConfigurationManager: FileConfigurationManager
+	) { }
+
+	async provideInlayHints(model: vscode.TextDocument, range: vscode.Range, token: vscode.CancellationToken): Promise<vscode.InlayHint[]> {
+		const filepath = this.client.toOpenedFilePath(model);
+		if (!filepath) {
+			return [];
+		}
+
+		await this.fileConfigurationManager.ensureConfigurationForDocument(model, token);
+
+		if (!this.someInlayHintsEnabled(model)) {
```

Comment:
Can we skip calling `ensureConfigurationForDocument` if inlay hints are disabled?

Reply:
`someInlayHintsEnabled` has been deleted. So I guess I don't need.